### ISSUE TITLE
Format Code not working

### ISSUE THREAD

**Issue by jrieken** [OmniSharp/omnisharp-vscode#660]
_From @johanneskopf on August 12, 2016 21:53_

- VSCode Version: 1.5.0-insider
- OS Version: Windows 10, Build 14901

I work with Unity, but no matter if I open a whole solution or only a .cs file: the formatting of the code with (Shift+Alt+F), or by using the action in the command palette, isn't working.

_Copied from original issue: Microsoft/vscode#10487_

**Comment by jrieken** [comment]
_From @brandonmp on August 13, 2016 0:7_

Will just add-on to this one though my circumstances are diff't:

VSCode v. 1.4.0
OS: Ubuntu 16.04

Has worked flawlessly for months, but just stopped yesterday. Uninstalled/reinstalled a few times, and tried in various projects/languages (namely: JS, JSX, and Python), still no formatting (either via shortcut or via command palette).

Admittedly I'm not sure how to -fully- uninstall the app, as every time I re-install, it remembers what files I had opened.  I imagine there are settings leftover somewhere, so I'll keep trying to find/delete those.


ETA: I was able to "fix" the problem finally:

1. Use `synaptic package manager` and mark the `code` package for 'Full Removal,' click `apply`.
2. Delete `~/.vscode`
3. Delete `~/.config`
4. Reinstall
5. Profit (Until it breaks again :( ... I can't go back to doing my own indentation! )

**Comment by jrieken** [comment]
_From @johanneskopf on August 13, 2016 19:52_

I tried to reinstall VS Code, deleting the extension folder and the VS Code Folder in APPDATA, but nothing changed.

**Comment by jrieken** [comment]
_From @brandonmp on August 13, 2016 21:15_

Does VSCode leave any windows registry keys?  Since it's a bug that starts spontaneously, I have to imagine there's some residual settings somewhere. 

If you install logged in as a  diff't user, any luck?

**Comment by johanneskopf** [comment]
Coincidentally today I saw that you have Developer Tools to enable.
When I enable them and press "Shift"+"Alt"+"F" to format my code, the warning "Ignoring commands" gets written. Maybe this has something to do why my format code is not working as expected?

So in cursor.ts, this if gets executed, therefore the warning gets written:
`var loserCursorsMap = this._getLoserCursorMap(rawOperations);
		if (loserCursorsMap.hasOwnProperty('0')) {
			// These commands are very messed up
			console.warn('Ignoring commands');
			return false;
		}`

**Comment by johanneskopf** [comment]
Is there some way I change the code of "cursor.ts" myself?
As it's path is somewhere in the internet, and therefore I can change this code and save it, but the new script doesn't get executed.

The path can be seen here:
![cursor_script_path](https://cloud.githubusercontent.com/assets/13490371/17750877/ee358946-64c4-11e6-8c90-52c14a17adee.PNG)

**Comment by johanneskopf** [comment]
Still happening with the latest 1.4 beta 8.
Any news here?

**Comment by DustinCampbell** [comment]
Hey @jrieken, based on @brandonmp's comments about this not working for him in JS, JSX and Python and @johanneskopf's investigation, I'm not sure this is a C# extension issue.

**Comment by DustinCampbell** [comment]
@johanneskopf: I see that the developer tools show the .NET CLI Tools message is appearing for you. I doubt it's related, but since you said you have 1.4.0-beta8 of the C# extension installed, you can disable that message (since you're not using the .NET CLI) by setting the following setting in VS Code:

```
csharp.suppressDotentInstallWarning: true
```

**Comment by johanneskopf** [comment]
@DustinCampbell I made a fresh install of VS-Code and it's plugins and upgraded to the 1.4.0-beta9 today, and then set your mentioned settings. I don't know why, but this setting isn't working for me:

![dotnet](https://cloud.githubusercontent.com/assets/13490371/18051778/eea54cfa-6df6-11e6-9e26-7ab8a5edc453.PNG)

I tried it inside the user and the workspace settings, but nothing changed.

**Comment by DustinCampbell** [comment]
@gregg-miskelly, @chuckries: Isn't the "csharp.suppressDotentInstallWarning" supposed to be working?

**Comment by gregg-miskelly** [comment]
Your string seems to have a typeo. It is `suppressDotnetInstallWarning`.

**Comment by DustinCampbell** [comment]
That was my typo. The screenshot that @johanneskopf provided above does not have the same typo.

**Comment by chuckries** [comment]
Something is a little screwy with @johanneskopf's screenshot. In versions of the extension where this can be disabled, we show a button in the alert to navigate to the user settings in the editor. 
![image](https://cloud.githubusercontent.com/assets/6598594/18060026/c6e6e1cc-6dd0-11e6-8d93-778ad77c7602.png)
It doesn't look like @johanneskopf's version of the extension has this button. The latest version of the extension may not actually be installed?

**Comment by johanneskopf** [comment]
@chuchries
I already recognized that no beta features are working. That's why I asked "What's the preferred way to install the latest beta on Windows?" Can you give me any advice here?

**Comment by DustinCampbell** [comment]
Sorry, I misread your reply @johanneskopt. It sounds like you don't actually have the beta installed. Installing the beta VSIX doesn't require moving source files around. That would generally just break the extension. Installing the extension is as easy as just selecting File->Open from the VS Code menu and pointing it at the VSIX. Could you try that?

**Comment by johanneskopf** [comment]
@DustinCampbell thank you! I was searching for an explanation like this. I don't know why I haven't tried that before.

The beta version is now working, but the warning that "Ignoring commands" persists. So format code is still not working.

![capture](https://cloud.githubusercontent.com/assets/13490371/18060802/53e0fe6a-6e20-11e6-95c6-baf4613f6e43.PNG)

**Comment by DustinCampbell** [comment]
OK. In that case, @jrieken, do you have any ideas? It's not clear to me how the C# Extension would contribute to this issue.

**Comment by jrieken** [comment]
The formatting providers are spec'd to produce an array of text edits where a [text edit](https://github.com/Microsoft/vscode-typescript) is a range (of existing text) and new text. There are cases in which we cannot process them and drop them - that's usually when they overlap or selecting something non existent. 

For a better I need exact step, esp current text and formatting options.

**Comment by DustinCampbell** [comment]
@jrieken, is the "Ignoring commands" message that @johanneskopf mentioned relevant? It seems rather ominous to me and I was guessing that <kbd>Shift+Alt+F</kbd> wasn't working because the editor was already in some sort of state where it would ignore it.

**Comment by johanneskopf** [comment]
Got another weird behavior today. Currently I have 1.4.0 of the C# extension and the latest insider build of VS Code installed.

Step by step:

1. Created a new Unity Project
2. Imported the "VS Code" Asset from the asset store
3. Opened preferences - VSCode and pressed "Write Workspace Settings"
4. Created a new script called "Test.cs"
5. Opened it by double clicking it in Unity
6. Wrote in the Start() function "int h=0;" -> should be corrected to "int h = 0;"
7. Pressed `Shift+Alt+F`

Before using the formating command:
![beforeformat](https://cloud.githubusercontent.com/assets/13490371/18099624/b77da258-6ee7-11e6-87a3-465f3a7b619a.PNG)

Afterwards:
![afterformat](https://cloud.githubusercontent.com/assets/13490371/18099630/be2ca644-6ee7-11e6-8741-336907a85494.PNG)

**Comment by johanneskopf** [comment]
@jrieken Anything else I can do to help you fix it?

**Comment by jrieken** [comment]
Yes - please attach the project as zip here

**Comment by nfantone** [comment]
Same behaviour here. Using latest VSCode`1.5.0-inside` on Mac OSX 10.11.6. Formatting either not working or just randomly scrambling letters (like [shown above](https://github.com/OmniSharp/omnisharp-vscode/issues/660#issuecomment-243515901)).

@jrieken Full project can be found [here](https://dl.dropboxusercontent.com/u/29835432/2D%20Roguelike.zip) (file larger > 10MB - couldn't attach it here).

**Comment by nfantone** [comment]
Ok, incidentally, I want to point out some other things that might come in handy when working out a solution for the bug:

- Creating a new empty project with `$ dotnet new` and trying to format code doesn't work very well either.
- I can say It's not Unity related.
- Functionality doesn't seem to be very consistent. In my tests, code formatting would sometimes work fine, don't work at all of mess things up. See screenshot below:

<img width="697" alt="screen shot 2016-09-05 at 2 03 54 pm" src="https://cloud.githubusercontent.com/assets/476069/18248676/9e2b9e52-7371-11e6-9c6e-4e17bea2e4c2.png">

Intuitively, it looks like a parsing issue. I managed to reproduce the "scrambling" problem following these steps:

1.  Create a new project with `$ dotnet new`.
2. Open the directory as a project in VSCode.
3. Restore project dependencies and let it install and fire up Omnisharp.
4. Open `Program.cs`.
5. At the beginning of _line 5_, delete the tab space.
6. Move curly brace from _line 4_ to the line above and put a space between `ConsoleApplication` and the brace. File should look like this:

<img width="494" alt="screen shot 2016-09-05 at 2 22 30 pm" src="https://cloud.githubusercontent.com/assets/476069/18249110/35198516-7374-11e6-9180-a9b52bf4a1e2.png">

Finally, select _Format code_ from the dropdown menu. Notice what it does to the code.


<img width="496" alt="screen shot 2016-09-05 at 2 19 00 pm" src="https://cloud.githubusercontent.com/assets/476069/18249031/baa3dfb6-7373-11e6-8fb0-a93a786f53c1.png">

**Comment by jrieken** [comment]
Thanks for the steps! This is actually a really bad regression from a change we made in August. Our intent was to fix the problem that many formatters basically replace the whole document text instead of producing 'true' and 'minimal' text edits. We added some post-processing that minimises the text edits and it seems that messes things up. 

I have created https://github.com/Microsoft/vscode/issues/11593 to track the issue and will make sure we ship the fix soon. Sorry, for not thinking about this earlier.
